Feature/updater

blackshadev opened the pull request:
Updater!

How to test:
- pull branch
- set .env vars
  - `UPDATER_GIT_SECRET=secret`
  - `UPDATER_GIT_BRANCH=feature/updater`
- run `artisan config:clear` and `artisan optimize:clear`
- checkout to previous commit (dont worry, it is just a newline removal) `git checkout HEAD^`
- goto `http://aas2.local/updater/version` see 61dc0ba
- goto `http://aas2.local/updater/update?secret=secret` wait and see output with b74f9c9
- goto `http://aas2.local/updater/version` see b74f9c9

robinklaassen commented:
Ik zou nog even de benodigde `env` vars toevoegen aan `.env.example`

robinklaassen commented on packages/Updater/Http/Controllers/UpdateController.php:
```diff
@@ -0,0 +1,43 @@
+        $this->updaterService->update();
+
+        return response()->json([
+            'success' => true,
```
Dus het gaat volgens de response altijd goed, ook als het fout gaat? 🤔

robinklaassen replied:
Oh nee dan wordt er een exceptie gegooid en krijg je dit bericht niet te zien

blackshadev replied:
Ja die success kan er ook uit.

robinklaassen commented on packages/Updater/Services/DateTimeProvider.php:
```diff
@@ -0,0 +1,11 @@
+
+namespace Updater\Services;
+
+class DateTimeProvider implements DateTimeProviderInterface
+{
+    public function now(): \DateTimeImmutable
+    {
+        return new \DateTimeImmutable();
+    }
+}
```
Hier een aparte class met interface voor maken voelt wel een beetje als overengineering

blackshadev replied:
Hiermee kan ik later makkelijker unit tests toevoegen

robinklaassen commented on packages/Updater/Services/Updater/UpdaterService.php:
```diff
@@ -0,0 +1,64 @@
+
+    protected function postUpdate()
+    {
+        PostUpdateEvent::dispatch();
```
Gebeurt het afhandelen van het event synchroon of asynchroon?

blackshadev replied:
synchroon

robinklaassen commented on packages/Updater/UpdaterServiceProvider.php:
```diff
@@ -0,0 +1,115 @@
+class UpdaterServiceProvider extends ServiceProvider
+{
+    protected $listen = [
+        PostUpdateEvent::class => [
+            ComposerPostUpdateListener::class,
+            ArtisanPostUpdateListener::class,
+        ],
```
Ligt de volgordelijkheid van de listeners hiermee vast of kan deze nog verschillen per fired event?

robinklaassen commented on packages/Updater/UpdaterServiceProvider.php:
```diff
@@ -0,0 +1,115 @@
+
+        $this->bindDependancies();
+
+        Event::listen(PostUpdateEvent::class, ComposerPostUpdateListener::class);
+        Event::listen(PreUpdateEvent::class, ArtisanPreUpdateListener::class);
+        Event::listen(PostUpdateEvent::class, ArtisanPostUpdateListener::class);
```
Hoe verhouden deze regels zich tot de `protected $listen` definitie bovenin de class?

blackshadev commented:
> Ik zou nog even de benodigde `env` vars toevoegen aan `.env.example`

Added UPDATER_SECRET (and renamed UPDATER_GIT_SECRET to UPDATER_SECRET) 
The rest is not needed on production (or normally)

robinklaassen reviewed (approved):
Ik begrijp nog niet hoe alles werkt, met name die service provider gaat wel erg diep het laravel framework in. Maar het werkt en is netjes opgezet.
